File: engine/b/src/main/java/net/crazyinvestor/engine_b/dto/request/BithumbWSSubscribeRequest.java
package net.crazyinvestor.engine_b.dto.request;

import lombok.Builder;
import net.crazyinvestor.engine_b.enums.BithumbOpType;
import net.crazyinvestor.engine_b.enums.BithumbSymbol;
import net.crazyinvestor.engine_b.enums.BithumbTickTypes;

import java.io.Serial;
import java.io.Serializable;
import java.util.List;

@Builder
public final class BithumbWSSubscribeRequest implements Serializable {
    @Serial
    private static final long serialVersionUID = -687691491344005033L;

    private final BithumbOpType opType;
    private final List<BithumbSymbol> symbols;
    private final List<BithumbTickTypes> tickTypes;

    @Override
    public String toString() {
        List<String> symbolStrList = this.symbols
                .stream()
                .map(BithumbSymbol::getType)
                .distinct() // 중복 제거
                .map(v -> "\"" + v + "\"") // 문자열 가공
                .toList(); // 리스트화

        List<String> tickTypeStrList = this.tickTypes
                .stream()
                .map(BithumbTickTypes::getTickType)
                .distinct() // 중복 제거
                .map(v -> "\"" + v + "\"") // 문자열 가공
                .toList(); // 리스트화

        return "{" +
                "\"type\" : \"" + this.opType.getOpType() + "\"" + ", " +
                "\"symbols\" : " + "[" + String.join(", ", symbolStrList) + "]" + ", " +
                "\"tickTypes\" : " + "[" + String.join(", ", tickTypeStrList) + "]" +
                "}";
    }
}

File: engine/a/ticker_producer_engine/src/main/kotlin/net/crazyinvestor/ticker_producer_engine/dto/request/BithumbWSSubscribeRequest.kt
package net.crazyinvestor.ticker_producer_engine.dto.request

import net.crazyinvestor.ticker_producer_engine.enums.BithumbOpType
import net.crazyinvestor.ticker_producer_engine.enums.BithumbSymbol
import net.crazyinvestor.ticker_producer_engine.enums.BithumbTickTypes
import java.io.Serializable

class BithumbWSSubscribeRequest(
    private val opType: BithumbOpType,
    private val symbols: List<BithumbSymbol>,
    private val tickTypes: List<BithumbTickTypes>
) : Serializable {
    override fun toString(): String {
        fun String.quoted(): String = "\"${this}\""

        val joinedStringForSymbols: String = symbols
            .map(BithumbSymbol::type)
            .distinct()
            .joinToString(transform = String::quoted)

        val joinedStringForTickTypes: String = tickTypes
            .map(BithumbTickTypes::tickType)
            .distinct()
            .joinToString(transform = String::quoted)

        return "{ ${"type".quoted()} : ${opType.opType.quoted()}, ${"symbols".quoted()} : [$joinedStringForSymbols], ${"tickTypes".quoted()} : [$joinedStringForTickTypes]}"
    }

    companion object {
        fun fromOpType(opType: BithumbOpType, symbols: List<BithumbSymbol>, tickTypes: List<BithumbTickTypes>): BithumbWSSubscribeRequest {
            return BithumbWSSubscribeRequest(
                opType = opType,
                symbols = symbols,
                tickTypes = tickTypes
            )
        }
    }
}